src/internal/AnalyticsContext.ts

```diff
@@ -0,0 +1,50 @@
+const deepmerge = require('deepmerge');
+import {AnalyticsEventModel} from './AnalyticsEventModel';
+import {DataMap, PromiseOrValue} from './types';
+
+export type AnalyticsFilter = (model: AnalyticsEventModel) => PromiseOrValue<void>;
+
+export class AnalyticsContext {
+  readonly Scopes: string[] = [];
+  readonly ExtraData: DataMap = {};
+  readonly MetaData: DataMap = {};
+  readonly Identities: DataMap = {};
+  readonly Filters: AnalyticsFilter[] = [];
+
+  union(analyticsContext?: AnalyticsContext): AnalyticsContext {
+    if (!analyticsContext) {
+      return this;
+    }
+
+    const union = new AnalyticsContext();
```

Comment:
This bit of code is clever, but hard to read and throws type safety out the window.
Feels to me like this was far easier to understand and just as good: https://github.com/Soluto/shisell-js/blob/master/lib/AnalyticsDispatcher.js#L13